src/main/webapp/WEB-INF/pages/recentlyViewedProducts.jsp

```diff
@@ -0,0 +1,30 @@
+<c:if test="${not empty sessionScope.recentlyViewedProducts}">
+    <div>
+      <h3>Recently Viewed Products</h3>
+      <table>
+        <tr>
+          <c:forEach var="recentProduct" items="${sessionScope.recentlyViewedProducts}">
+            <td>
+              <img src="${recentProduct.imageUrl}">
+            </td>
+          </c:forEach>
+        </tr>
+        <tr>
+          <c:forEach var="recentProduct" items="${sessionScope.recentlyViewedProducts}">
+            <td>
+              <a href="${pageContext.servletContext.contextPath}/products/${recentProduct.id}">
+                ${recentProduct.description}
+              </a>
+            </td>
+            </c:forEach>
+        </tr>
+        <tr>
+          <c:forEach var="recentProduct" items="${sessionScope.recentlyViewedProducts}">
```

Comment:
when we call some code few times (sessionScope.recentlyViewedProducts in your example), then it makes sense to create a variable and save this object to variable